File: src/Algorithms/mazeGenerator.js
export function generateMaze(numRows, numCols, singlePath) {
  let maze = new Maze(numRows, numCols);

  // Generate the maze
  while (maze.stack.length > 0 || !maze.currentCell.visited) {
    maze.generateMazeNotGlobal();
  }

  if (!singlePath) {
    for (let i = 0; i < numRows * numCols * 0.1; i++) { // remove 10% of the walls
      let x = Math.floor(Math.random() * (numRows - 2)) + 1;
      let y = Math.floor(Math.random() * (numCols - 2)) + 1;
      maze.grid[x][y].isWall = false;
    }
  }

  // Convert the maze grid to the format expected by the rest of your application
  let grid1 = maze.grid.map(row => row.map(cell => createNode(cell.x, cell.y, cell.isWall, maze.start, maze.end, 1)));
  let grid2 = maze.grid.map(row => row.map(cell => createNode(cell.x, cell.y, cell.isWall, maze.start, maze.end, 2)));
  let grid3 = maze.grid.map(row => row.map(cell => createNode(cell.x, cell.y, cell.isWall, maze.start, maze.end, 3)));
  let grid4 = maze.grid.map(row => row.map(cell => createNode(cell.x, cell.y, cell.isWall, maze.start, maze.end, 4)));
  let grid5 = maze.grid.map(row => row.map(cell => createNode(cell.x, cell.y, cell.isWall, maze.start, maze.end, 5)));

  // Find start and end nodes
  let gridDijkstraStartNode = grid1[maze.start.y][maze.start.x];
  let gridDijkstraEndNode = grid1[maze.end.y][maze.end.x];
  let gridAstarStartNode = grid2[maze.start.y][maze.start.x];
  let gridAstarEndNode = grid2[maze.end.y][maze.end.x];
  let gridBfsStartNode = grid3[maze.start.y][maze.start.x];
  let gridBfsEndNode = grid3[maze.end.y][maze.end.x];
  let gridDfsStartNode = grid4[maze.start.y][maze.start.x];
  let gridDfsEndNode = grid4[maze.end.y][maze.end.x];
  let gridWallFollowerStartNode = grid5[maze.start.y][maze.start.x];
  let gridWallFollowerEndNode = grid5[maze.end.y][maze.end.x];


  return {
    gridDijsktra: grid1,
    gridAstar: grid2,
    gridBfs: grid3,
    gridDfs: grid4,
    gridWallFollower: grid5,
    gridDijkstraStartNode,
    gridDijkstraEndNode,
    gridAstarStartNode,
    gridAstarEndNode,
    gridBfsStartNode,
    gridBfsEndNode,
    gridDfsStartNode,
    gridDfsEndNode,
    gridWallFollowerStartNode,
    gridWallFollowerEndNode
  };
}

class Cell {
  constructor(x, y, isWall = false) {
    this.x = x;
    this.y = y;
    this.isWall = isWall;
    this.visited = false;
  }
}

class Maze {
  constructor(width, height) {
    if (width % 2 ==0 ){
      width++;
    }
    if (height % 2 == 0){
      height++;
    }
    this.width = width
    this.height = height
    this.grid = [];
    this.stack = [];

    for (let y = 0; y < this.height; y++) {
      let row = [];
      for (let x = 0; x < this.width; x++) {
        let isWall = x % 2 === 0 || y % 2 === 0; // All cells at even indices are walls
        row.push(new Cell(x, y, isWall));
      }
      this.grid.push(row);
    }

    this.currentCell = this.grid[1][1]; // Start at the top-left corner
    this.start = this.grid[1][1]; // Top-left corner
    this.end = this.grid[this.height - 2][this.width - 2]; // Bottom-right corner
  }

  getCell(x, y) {
    if (x < 0 || y < 0 || x >= this.width || y >= this.height) {
      return undefined;
    }
    return this.grid[y][x];
  }

  getNeighbors(cell) {
    let neighbors = [];

    // Get the cells above, below, to the left, and to the right of the given cell
    let top = this.getCell(cell.x, cell.y - 2);
    let right = this.getCell(cell.x + 2, cell.y);
    let bottom = this.getCell(cell.x, cell.y + 2);
    let left = this.getCell(cell.x - 2, cell.y);

    // If a cell is within the grid and has not been visited yet, add it to the list of neighbors
    if (top && !top.visited) neighbors.push(top);
    if (right && !right.visited) neighbors.push(right);
    if (bottom && !bottom.visited) neighbors.push(bottom);
    if (left && !left.visited) neighbors.push(left);

    if (neighbors.length > 0) {
      // 75% chance to return a random neighbor
      if (Math.random() < 0.75) {
        return neighbors[Math.floor(Math.random() * neighbors.length)];
      }

      // Otherwise, sort the neighbors based on their distance from the start point
      neighbors.sort((a, b) => {
        let aDistance = Math.hypot(a.x - this.start.x, a.y - this.start.y);
        let bDistance = Math.hypot(b.x - this.start.x, b.y - this.start.y);
        return bDistance - aDistance;
      });

      // Return the neighbor that is furthest from the start point
      return neighbors[0];
    } else {
      return undefined;
    }
  }

  generateMazeNotGlobal() {
    this.currentCell.visited = true;
    let nextCell = this.getNeighbors(this.currentCell);

    if (nextCell) {
      nextCell.visited = true;

      this.stack.push(this.currentCell);

      // Remove the wall between the current cell and the next cell
      let wallX = (this.currentCell.x + nextCell.x) / 2;
      let wallY = (this.currentCell.y + nextCell.y) / 2;
      this.grid[wallY][wallX].isWall = false;

      this.currentCell = nextCell;
    } else if (this.stack.length > 0) {
      // 40% chance to backtrack and carve a new path
      if (Math.random() < 0.4) {
        let backtrackCell = this.stack[Math.floor(Math.random() * this.stack.length)];
        this.currentCell = backtrackCell;
      } else {
        this.currentCell = this.stack.pop();
      }
    }
  }

}

function createNode(col, row, isWall, startNode, endNode, gridId) {
  return {
    col,
    row,
    isStart: startNode && col === startNode.x && row === startNode.y,
    isEnd: endNode && col === endNode.x && row === endNode.y,
    distance: Infinity,
    isVisited: false,
    isWall,
    previousNode: null,
    gridId,
  };
}

File: golang-src/maze/maze_generator.go
package maze

import (
	"math"
	"math/rand"
	"time"
)

type Cell struct {
	X, Y    uint16
	IsWall  bool
	Visited bool
}

type Node struct {
	X, Y         uint16
	IsStart      bool
	IsEnd        bool
	Distance     uint32
	IsVisited    bool
	IsWall       bool
	PreviousNode *Node
	GridId       uint8
	NoOfVisits   uint8
	H, F         float32
}

type Maze struct {
	Width, Height int
	Grid          [][]Cell
	Stack         []*Cell
	CurrentCell   *Cell
	Start         *Cell
	End           *Cell
}

func NewMaze(width, height int) *Maze {
	// Increase dimensions by 1 if they're even
	if width%2 == 0 {
		width++
	}
	if height%2 == 0 {
		height++
	}

	m := &Maze{
		Width:  width,
		Height: height,
		Grid:   make([][]Cell, height),
		Stack:  make([]*Cell, 0, (width*height)/2), // Preallocate stack with estimated capacity
	}

	// Initialize the grid with walls and paths
	for y := 0; y < height; y++ {
		m.Grid[y] = make([]Cell, width)
		for x := 0; x < width; x++ {
			isWall := (x%2 == 0) || (y%2 == 0)
			m.Grid[y][x] = Cell{X: uint16(x), Y: uint16(y), IsWall: isWall}
		}
	}

	// Set start and end cells
	m.CurrentCell = &m.Grid[1][1]
	m.Start = &m.Grid[1][1]
	m.End = &m.Grid[height-2][width-2]

	// Ensure start and end cells are not walls
	m.Grid[1][1].IsWall = false
	m.Grid[height-2][width-2].IsWall = false

	return m
}

func (m *Maze) getCell(x, y int) *Cell {
	if x < 0 || y < 0 || x >= m.Width || y >= m.Height {
		return nil
	}
	return &m.Grid[y][x]
}

func (m *Maze) getNeighbors(cell *Cell) []*Cell {
	var neighbors []*Cell

	top := m.getCell(int(cell.X), int(cell.Y)-2)
	right := m.getCell(int(cell.X)+2, int(cell.Y))
	bottom := m.getCell(int(cell.X), int(cell.Y)+2)
	left := m.getCell(int(cell.X)-2, int(cell.Y))

	if top != nil && !top.Visited {
		neighbors = append(neighbors, top)
	}
	if right != nil && !right.Visited {
		neighbors = append(neighbors, right)
	}
	if bottom != nil && !bottom.Visited {
		neighbors = append(neighbors, bottom)
	}
	if left != nil && !left.Visited {
		neighbors = append(neighbors, left)
	}

	return neighbors
}

func (m *Maze) generateMazeNotGlobal() {
	m.CurrentCell.Visited = true
	nextCell := m.getNeighbors(m.CurrentCell)

	if nextCell != nil {
		nextCellCell := nextCell[rand.Intn(len(nextCell))] // Choose a random neighbor
		nextCellCell.Visited = true
		m.Stack = append(m.Stack, m.CurrentCell)

		// Calculate the wall position correctly
		wallX := (int(m.CurrentCell.X) + int(nextCellCell.X)) / 2
		wallY := (int(m.CurrentCell.Y) + int(nextCellCell.Y)) / 2

		if wallX >= 0 && wallX < m.Width && wallY >= 0 && wallY < m.Height {
			m.Grid[wallY][wallX].IsWall = false
		}

		m.CurrentCell = nextCellCell
	} else if len(m.Stack) > 0 {
		if rand.Float64() < 0.4 {
			backtrackCell := m.Stack[rand.Intn(len(m.Stack))]
			m.CurrentCell = backtrackCell
		} else {
			m.CurrentCell = m.Stack[len(m.Stack)-1]
			m.Stack = m.Stack[:len(m.Stack)-1]
		}
	}
}

func createNode(x, y uint16, isWall bool, start, end *Cell, gridId uint8) Node {
	return Node{
		X:            x,
		Y:            y,
		IsStart:      start != nil && x == start.X && y == start.Y,
		IsEnd:        end != nil && x == end.X && y == end.Y,
		Distance:     math.MaxUint32,
		IsVisited:    false,
		IsWall:       isWall,
		PreviousNode: nil,
		GridId:       gridId,
	}
}

func GenerateMaze(numRows, numCols int, singlePath bool) map[string]interface{} {
	s := rand.NewSource(time.Now().UnixNano())
	r := rand.New(s)
	maze := NewMaze(numCols, numRows) // Note: numCols is width, numRows is height

	// Generate maze with the current implementation
	for len(maze.Stack) > 0 || !maze.CurrentCell.Visited {
		maze.generateMazeNotGlobal()
	}

	// Add extra paths if not single path
	if !singlePath {
		for i := 0; i < numRows*numCols/10; i++ {
			x := r.Intn(numCols)
			y := r.Intn(numRows)
			if x > 0 && x < numCols-1 && y > 0 && y < numRows-1 {
				maze.Grid[y][x].IsWall = false
			}
		}
	}

	// Generate grids for different algorithms
	grids := make(map[int][][]Node, 5)
	for i := 1; i <= 5; i++ {
		grid := make([][]Node, len(maze.Grid))
		for y, row := range maze.Grid {
			grid[y] = make([]Node, len(row))
			for x, cell := range row {
				grid[y][x] = createNode(cell.X, cell.Y, cell.IsWall, maze.Start, maze.End, uint8(i))
			}
		}
		grids[i] = grid
	}

	// Return maze data for different algorithms
	return map[string]interface{}{
		"gridDijkstra":              grids[1],
		"gridAstar":                 grids[2],
		"gridBFS":                   grids[3],
		"gridDFS":                   grids[4],
		"gridWallFollower":          grids[5],
		"gridDijkstraStartNode":     &grids[1][maze.Start.Y][maze.Start.X],
		"gridDijkstraEndNode":       &grids[1][maze.End.Y][maze.End.X],
		"gridAstarStartNode":        &grids[2][maze.Start.Y][maze.Start.X],
		"gridAstarEndNode":          &grids[2][maze.End.Y][maze.End.X],
		"gridBFSStartNode":          &grids[3][maze.Start.Y][maze.Start.X],
		"gridBFSEndNode":            &grids[3][maze.End.Y][maze.End.X],
		"gridDFSStartNode":          &grids[4][maze.Start.Y][maze.Start.X],
		"gridDFSEndNode":            &grids[4][maze.End.Y][maze.End.X],
		"gridWallFollowerStartNode": &grids[5][maze.Start.Y][maze.Start.X],
		"gridWallFollowerEndNode":   &grids[5][maze.End.Y][maze.End.X],
	}
}
